mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around and cause us to lose data due to
+/// variable sized data being written. If W + data size > R and should write
+/// precede read, the data could become corrupted.
+///
+/// Either this is prevented by scanning the file for the next block and
+/// checking if the write will fit. Or, W + data must not pass both D and R.
+/// - Not passing D helps with guaranteeing all data read was handled.
+/// - Not passing R helps avoid corruption.
+#[allow(dead_code)]
+pub(crate) struct RingBuffer<'a> {
+    flush_options: FlushOptions,
+    flush_state: FlushState,
+    mmap: &'a mut MmapMut,
+    max_file_size: usize,
+    order: usize,
+    pointers: FilePointers,
+    waker: Option<Waker>,
+}
+
+#[allow(dead_code)]
+impl<'a> RingBuffer<'a> {
+    fn new(mmap: &'a mut MmapMut, max_file_size: usize, flush_options: FlushOptions) -> Self {
+        let file_pointers =
+            find_pointers_post_crash(&mmap, max_file_size).expect("Failed to init file pointers");
+        Self {
+            flush_options,
+            flush_state: FlushState::new(),
+            mmap,
+            max_file_size,
+            order: usize::default(),
+            pointers: file_pointers,
+            waker: None,
+        }
+    }
+
+    fn insert(&mut self, publication: &Publication) -> StorageResult<Key> {
+        let timer = Instant::now();
+        let data = binary_serialize(publication)?;
+        let data = Data::new(&data);
+        let data_size = binary_serialize_size(&data)?;
+        let data_hash = calculate_hash(&data);
+
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read;
+        let order = self.order;
+        let key = write_index;
+
+        let block_header =
+            BlockHeaderWithHash::new(data_hash, data_size, order, read_index, write_index);
+        let block_size = serialized_block_size()?;
+
+        let total_size = block_size + data_size;
+        // Check to see if we might corrupt data if we write, if so, return err of full.
+        // There are two cases for this:
+        // 1. write has wrapped around and is now behind read
+        if write_index < read_index && write_index + total_size > read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::Full));
+        }
+        // 2. write has reached the end (or close enough) but read has not moved
+        if read_index == 0 && write_index + total_size > self.max_file_size {
+            return Err(StorageError::RingBuffer(RingBufferError::Full));
+        }
+
+        let mut start = write_index;
+
+        // Check if an existing block header is present. If the block is there
+        // and if the overwrite flag is not set then we shouldn't write.
+        let result = load_block_header(self.mmap, start, block_size, self.max_file_size);
```

Comment:
There are several possibilities: 
1. There is no data, this will deserialize as an empty block
2. There is a block and we deserialize
3. There may have been a block but the data is not deserializable into a block, which means this data would be lost. We have an option of trying to do a partial deserialize and hope to recover the data... but if the block is corrupt i.e. non deserializable... then seems it likely lost cause.

So, the reason I didn't care about err here is it seemed to be only possible if it was too late to do anything about. If we add versioning or different blocks this will 100% need to be handled.

Reply:
ok